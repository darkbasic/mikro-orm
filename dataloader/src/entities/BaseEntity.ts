import {IdEntity, PrimaryKey} from 'mikro-orm';

interface BaseEntityConstructor {
  id?: number;
}

export abstract class BaseEntity implements IdEntity<BaseEntity> {
  @PrimaryKey()
  id!: number;

  constructor({id}: BaseEntityConstructor = {}) {
    if (id) {
      this.id = id;
    }
  }
}
