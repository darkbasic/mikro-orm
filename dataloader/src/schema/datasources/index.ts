export * from './OrmAPI';
export * from './UserAPI';
export * from './SportAPI';
export * from './SpecialtyAPI';
export * from './SiteAPI';
export * from './MatchAPI';
