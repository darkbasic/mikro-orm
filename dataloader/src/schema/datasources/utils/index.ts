import {GraphQLResolveInfo} from 'graphql';
import {
  Reference,
  AnyEntity,
  EntityManager,
  Primary,
  IdentifiedReference,
  Collection,
  WrappedEntity,
  wrap,
  EntityRepository,
} from 'mikro-orm';
import DataLoader from 'dataloader';

export function ensureIsNum(id: string | number): number;
export function ensureIsNum(
  id: string | number | null | undefined
): number | undefined;
export function ensureIsNum(ids: Array<string | number>): number[];
export function ensureIsNum(
  ids: Array<string | number> | null | undefined
): number[] | undefined;
export function ensureIsNum(
  idOrIds: null | undefined | string | number | (string | number)[]
): undefined | number | number[] {
  if (idOrIds === null || idOrIds === undefined) {
    return undefined;
  }
  if (idOrIds instanceof Array) {
    return idOrIds.map(id => {
      const num = Number(id);
      if (isNaN(num)) {
        throw new Error(`The ID ${id} is not a valid number`);
      }
      return num;
    });
  } else {
    const num = Number(idOrIds);
    if (isNaN(num)) {
      throw new Error(`The ID ${idOrIds} is not a valid number`);
    }
    return num;
  }
}

export function shouldPopulate({fieldNodes}: GraphQLResolveInfo): boolean {
  if (fieldNodes.length === 1) {
    const selections = fieldNodes[0].selectionSet?.selections;
    if (selections?.length === 1) {
      if (selections[0].kind !== 'InlineFragment') {
        return selections[0].name.value !== 'id';
      }
    }
  }
  return true;
}

export function getItemNoPopulate<T>(
  ref: Reference<T>,
  info?: GraphQLResolveInfo
): T | undefined {
  return !info || shouldPopulate(info) ? undefined : ((ref as unknown) as T);
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function getAllPropertyNames(obj: any) {
  const props: string[] = [];

  do {
    Object.getOwnPropertyNames(obj).forEach(prop => {
      if (props.indexOf(prop) === -1) {
        props.push(prop);
      }
    });
  } while ((obj = Object.getPrototypeOf(obj)));

  return props;
}

interface Id {
  id: number | string;
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
function hasRef<T>(
  entity: T
): T &
  WrappedEntity<T, keyof T> &
  Record<string, IdentifiedReference<AnyEntity>> {
  return entity as T &
    WrappedEntity<T, keyof T> &
    Record<string, IdentifiedReference<AnyEntity>>;
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
function hasCol<T>(
  entity: T
): T & WrappedEntity<T, keyof T> & Record<string, Collection<AnyEntity>> {
  return entity as T &
    WrappedEntity<T, keyof T> &
    Record<string, Collection<AnyEntity>>;
}

function hasRefOrCol<T>(
  entity: T
): T &
  WrappedEntity<T, keyof T> &
  Record<string, IdentifiedReference<AnyEntity> | Collection<AnyEntity>> {
  return entity as T &
    WrappedEntity<T, keyof T> &
    Record<string, IdentifiedReference<AnyEntity> | Collection<AnyEntity>>;
}

function isRef<T>(
  refOrCol: IdentifiedReference<T> | Collection<T>
): refOrCol is IdentifiedReference<T> {
  return !(refOrCol instanceof Collection);
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
function isCol<T>(
  refOrCol: IdentifiedReference<T> | Collection<T>
): refOrCol is Collection<T> {
  return refOrCol instanceof Collection;
}

type RepoFind<K> = {
  repo: EntityRepository<K>;
  filter: Record<string, Primary<K> | Primary<K>[]>;
  many: boolean;
};

type Result<K> = {
  entityName: string;
  keys: string[];
  entitiesOrError: K[] | Error;
};

function toArray<T>(arrOrAny: T | T[]): T[] {
  return Array.isArray(arrOrAny) ? arrOrAny : [arrOrAny];
}

function objectsHaveSameKeys<T>(
  ...objects: Record<string, Primary<T> | Primary<T>[]>[]
): boolean {
  const allKeys = new Set(
    objects.reduce(
      (keys, object) => keys.concat(Object.keys(object)),
      [] as string[]
    )
  );
  return objects.every(object => allKeys.size === Object.keys(object).length);
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function arrayAreEqual<T extends Primary<any>[] | string[]>(
  a: T,
  b: T
): boolean {
  return JSON.stringify(a.sort()) === JSON.stringify(b.sort());
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function arrayIncludes<T extends (string | Primary<any>)[]>(
  arr: T,
  target: T
): boolean {
  return target.every(v => arr.includes(v));
}

export function groupPrimaryKeysByEntity<T extends AnyEntity<T>>(
  refs: readonly Reference<T>[]
) {
  return refs.reduce((acc, cur) => {
    const {className} = cur.__meta;
    const primaryKeys: Primary<T>[] = acc[className];
    const primaryKey: Primary<T> = cur.__primaryKey;
    acc[className] = primaryKeys
      ? [...new Set([...primaryKeys, primaryKey])]
      : [primaryKey];
    return acc;
  }, {} as Record<string, Primary<T>[]>);
}

export function groupInversedOrMappedKeysByEntity<T extends AnyEntity<T>>(
  collections: readonly Collection<T>[]
): Record<string, Record<string, Primary<T>[]>> {
  return collections.reduce((acc, cur) => {
    const className = cur.property.type;
    const record = acc[className] || (acc[className] = {});
    // Many to Many vs One to Many
    const inversedOrMappedBy = cur.property.inversedBy
      ? 'inversedBy'
      : cur.property.mappedBy
      ? 'mappedBy'
      : undefined;
    if (!inversedOrMappedBy) {
      throw new Error('Cannot find inversedBy or mappedBy prop');
    }
    const key: Primary<T> = cur.owner.__primaryKey;
    const keys: Primary<T>[] = record[cur.property[inversedOrMappedBy]];
    record[cur.property[inversedOrMappedBy]] = keys
      ? [...new Set([...keys, key])]
      : [key];
    return acc;
  }, {} as Record<string, Record<string, Primary<T>[]>>);
}

export function groupFindQueries<T extends AnyEntity<T>>(
  repoFinds: readonly RepoFind<T>[]
) {
  return repoFinds.reduce((acc, {repo, filter}) => {
    const entityName = repo['entityName'] as string;
    if (acc[entityName]) {
      let matchFound = false;
      const curFilters = acc[entityName].map(el => ({...el}));
      for (const curFilter of curFilters) {
        if (objectsHaveSameKeys(curFilter, filter)) {
          matchFound = true;
          Object.entries(filter).forEach(([prop, idOrIds]) => {
            const curIdOrIds = curFilter[prop];
            if (
              Array.isArray(curIdOrIds) ||
              Array.isArray(idOrIds) ||
              (curIdOrIds && curIdOrIds !== idOrIds)
            ) {
              curFilter[prop] = [
                ...new Set([...toArray(curIdOrIds ?? []), ...toArray(idOrIds)]),
              ];
            } else {
              curFilter[prop] = idOrIds;
            }
          });
          break;
        }
      }
      if (matchFound) {
        acc[entityName] = [...curFilters];
      } else {
        acc[entityName] = [...curFilters, filter];
      }
    } else {
      acc[entityName] = [filter];
    }
    return acc;
  }, {} as Record<string, Record<string, Primary<T> | Primary<T>[]>[]>);
}

export class EntityDataLoader<T extends Id = Id, K = AnyEntity<T, 'id'>> {
  private bypass: boolean;
  private refLoader: DataLoader<IdentifiedReference<K>, K>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private colLoader: DataLoader<Collection<any>, K[]>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private findLoader: DataLoader<RepoFind<any>, K[] | K | null>;

  constructor(em: EntityManager, bypass = false) {
    this.bypass = bypass;

    this.refLoader = new DataLoader<IdentifiedReference<K>, K>(async refs => {
      const groupedIds = groupPrimaryKeysByEntity(refs);
      const promises = Object.entries(groupedIds).map(([entity, ids]) =>
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        em.getRepository<K>(entity).find(ids as any)
      );
      await Promise.all(promises);
      return Promise.all(refs.map(async ref => await ref.load()));
    });

    this.colLoader = new DataLoader<Collection<K>, K[]>(async collections => {
      const groupedKeys = groupInversedOrMappedKeysByEntity(collections);
      const promises: Promise<K[]>[] = Object.entries(
        groupedKeys
      ).map(([entity, record]) =>
        em.getRepository<K>(entity).find(record, Object.keys(record))
      );
      const results = (await Promise.all(promises))
        .flat()
        .map(entity => wrap(entity));
      return collections.map(collection =>
        results.filter(result => {
          // Class matches
          if (result.__meta.className === collection.property.type) {
            const inversedOrMappedBy = collection.property.inversedBy
              ? 'inversedBy'
              : collection.property.mappedBy
              ? 'mappedBy'
              : undefined;
            if (!inversedOrMappedBy) {
              throw new Error('Cannot find inversedBy or mappedBy prop');
            }
            const refOrCol = hasRefOrCol(result)[
              collection.property[inversedOrMappedBy]
            ];
            return isRef(refOrCol)
              ? refOrCol.__primaryKey === collection.owner.__primaryKey
              : refOrCol
                  .getItems()
                  .map(el => el.__primaryKey)
                  .includes(collection.owner.__primaryKey);
          }
          return false;
        })
      );
    });

    this.findLoader = new DataLoader<RepoFind<K>, K[] | K | null>(
      async repoFinds => {
        const groupedKeys = groupFindQueries(repoFinds);
        const groupedResults: Result<K>[] = await Promise.all(
          Object.entries(groupedKeys)
            .map(([entityName, records]) => {
              return records.map(async record => {
                let entitiesOrError: K[] | Error;
                try {
                  entitiesOrError = await em
                    .getRepository<K>(entityName)
                    .find(record, Object.keys(record));
                } catch (e) {
                  entitiesOrError = e as Error;
                }
                return {
                  entityName,
                  keys: Object.keys(record),
                  entitiesOrError,
                };
              });
            })
            .flat()
        );

        return repoFinds.map(({repo, filter, many}) => {
          const result = groupedResults.find(
            ({entityName, keys, entitiesOrError}) => {
              if (
                entityName === String(repo['entityName']) &&
                arrayAreEqual(Object.keys(filter), keys)
              ) {
                if (!(entitiesOrError instanceof Error)) {
                  return true;
                } else {
                  throw entitiesOrError;
                }
              }
              return false;
            }
          );
          if (!result) {
            throw new Error('Cannot match results');
          }

          const {entitiesOrError} = result;
          if (!(entitiesOrError instanceof Error)) {
            return (
              entitiesOrError[many ? 'filter' : 'find'](entity => {
                for (const key of Object.keys(filter)) {
                  if (
                    !arrayIncludes(
                      toArray(filter[key]),
                      // eslint-disable-next-line @typescript-eslint/no-explicit-any
                      toArray((entity as any)[key]).map(
                        el => el.id as Primary<K>
                      )
                    )
                  ) {
                    return false;
                  }
                }
                return true;
              }) ?? null
            );
          } else {
            throw entitiesOrError;
          }
        });
      }
    );
  }

  load<L extends K>(ref: IdentifiedReference<L>, bypass?: boolean): Promise<L>;
  load<L extends K>(collection: Collection<L>, bypass?: boolean): Promise<L[]>;
  load<L extends K>(
    refOrCol: IdentifiedReference<L> | Collection<L>,
    bypass?: boolean
  ): Promise<L> | Promise<L[]> {
    if (isRef(refOrCol)) {
      return bypass ?? this.bypass
        ? refOrCol.load()
        : (this.refLoader.load(refOrCol) as Promise<L>);
    } else {
      return bypass ?? this.bypass
        ? refOrCol.loadItems()
        : (this.colLoader.load(refOrCol) as Promise<L[]>);
    }
  }

  find<L extends K>(
    repo: EntityRepository<L>,
    filter: Record<string, Primary<L> | Primary<L>[]>,
    bypass?: boolean
  ): Promise<L[]> {
    return bypass ?? this.bypass
      ? repo.find(filter)
      : (this.findLoader.load({repo, filter, many: true}) as Promise<L[]>);
  }

  findOne<L extends K>(
    repo: EntityRepository<L>,
    filter: Record<string, Primary<L> | Primary<L>[]>,
    bypass?: boolean
  ): Promise<L | null> {
    return bypass ?? this.bypass
      ? repo.findOne(filter)
      : (this.findLoader.load({
          repo,
          filter,
          many: false,
        }) as Promise<L | null>);
  }
}
