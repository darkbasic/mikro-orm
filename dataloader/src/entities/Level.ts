import {
  Entity,
  Property,
  ManyToOne,
  Unique,
  IdentifiedReference,
  Reference,
} from 'mikro-orm';
import {BaseEntity} from './BaseEntity';
import {User} from './User';
import {Sport} from './Sport';

interface LevelConstructor {
  user: User;
  sport: Sport;
  declaredLevel: number;
}

export {Level as LevelEntity};

@Entity()
@Unique({properties: ['user', 'sport']})
export class Level extends BaseEntity {
  @ManyToOne()
  user: IdentifiedReference<User>;

  @ManyToOne()
  sport: IdentifiedReference<Sport>;

  @Property({type: 'float'})
  declaredLevel: number;

  constructor({user, sport, declaredLevel}: LevelConstructor) {
    super();
    this.user = Reference.create(user);
    this.sport = Reference.create(sport);
    this.declaredLevel = declaredLevel;
  }
}
