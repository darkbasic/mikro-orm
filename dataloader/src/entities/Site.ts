import {Entity, Property, ManyToMany, Collection, OneToMany} from 'mikro-orm';
import {BaseEntity} from './BaseEntity';
import {Sport} from './Sport';
import {Match} from './Match';

interface SiteConstructor {
  id?: number;
  name: string;
  position: string;
  sports: Sport[];
}

export {Site as SiteEntity};

@Entity()
export class Site extends BaseEntity {
  @Property()
  name: string;

  @Property()
  position: string;

  @ManyToMany()
  sports: Collection<Sport> = new Collection<Sport>(this);

  @OneToMany(() => Match, match => match.site)
  matches: Collection<Match> = new Collection<Match>(this);

  constructor({id, name, position, sports}: SiteConstructor) {
    super({id});
    this.name = name;
    this.position = position;
    this.sports.add(...sports);
  }
}
