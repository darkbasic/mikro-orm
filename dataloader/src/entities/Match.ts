import {
  Entity,
  Property,
  ManyToMany,
  Collection,
  ManyToOne,
  IdentifiedReference,
  Reference,
} from 'mikro-orm';
import {BaseEntity} from './BaseEntity';
import {User} from './User';
import {Specialty} from './Specialty';
import {Site} from './Site';

interface MatchConstructor {
  date: Date;
  creator: User;
  specialty: Specialty;
  site: Site;
  partecipants: User[];
}

export {Match as MatchEntity};

@Entity()
export class Match extends BaseEntity {
  @Property()
  date: Date;

  @ManyToOne()
  creator: IdentifiedReference<User>;

  @ManyToOne()
  specialty: IdentifiedReference<Specialty>;

  @ManyToOne()
  site: IdentifiedReference<Site>;

  @ManyToMany()
  partecipants: Collection<User> = new Collection<User>(this);

  constructor({
    date,
    creator,
    specialty,
    site,
    partecipants,
  }: MatchConstructor) {
    super();
    this.date = date;
    this.creator = Reference.create(creator);
    this.specialty = Reference.create(specialty);
    this.site = Reference.create(site);
    this.partecipants.add(...partecipants);
  }
}
