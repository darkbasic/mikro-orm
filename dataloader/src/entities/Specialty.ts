import {
  Entity,
  Property,
  ManyToOne,
  OneToMany,
  Collection,
  IdentifiedReference,
  Reference,
} from 'mikro-orm';
import {BaseEntity} from './BaseEntity';
import {Sport} from './Sport';
import {Match} from './Match';

interface SpecialtyConstructor {
  id?: number;
  name: string;
  males: number;
  females: number;
  sport: Sport;
}

export {Specialty as SpecialtyEntity};

@Entity()
export class Specialty extends BaseEntity {
  @Property()
  name: string;

  @Property()
  males: number;

  @Property()
  females: number;

  @ManyToOne()
  sport: IdentifiedReference<Sport>;

  @OneToMany(() => Match, match => match.specialty)
  matches: Collection<Match> = new Collection<Match>(this);

  constructor({id, name, males, females, sport}: SpecialtyConstructor) {
    super({id});
    this.name = name;
    this.males = males;
    this.females = females;
    this.sport = Reference.create(sport);
  }
}
