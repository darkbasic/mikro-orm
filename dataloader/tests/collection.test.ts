import {MikroORM, EntityRepository} from 'mikro-orm';
import {PostgreSqlDriver} from 'mikro-orm/dist/drivers/PostgreSqlDriver';
import {initORMPostgreSql, resetDatabase} from './bootstrap';
import {EntityDataLoader} from '../src/schema/datasources/utils';
import {Match} from '../src/entities/Match';

function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

describe('Backend', () => {
  jest.setTimeout(10e3);
  let orm: MikroORM<PostgreSqlDriver>;
  let matchRepo: EntityRepository<Match>;
  let dataloader: EntityDataLoader;

  beforeAll(async () => {
    orm = await initORMPostgreSql();
    await resetDatabase(orm);
    matchRepo = orm.em.getRepository(Match);
    dataloader = new EntityDataLoader(orm.em);
  });

  beforeEach(async () => {
    orm.em.clear();
    await sleep(100);
  });

  test('isConnected()', async () => {
    await expect(orm.isConnected()).resolves.toBe(true);
    await orm.close(true);
    await expect(orm.isConnected()).resolves.toBe(false);
    await orm.connect();
    await expect(orm.isConnected()).resolves.toBe(true);
  });

  it('should retrieve parallel collections with EntityDataLoader', async () => {
    const matches = await matchRepo.find({id: {$lte: 5}});
    const partecipantsDataloader = await Promise.all(
      matches.map(({partecipants}) => dataloader.load(partecipants))
    );
    expect(partecipantsDataloader).toBeDefined();
    expect(partecipantsDataloader.length).toEqual(5);
    expect(partecipantsDataloader).toMatchSnapshot();
  });

  it('should return the same entities without EntityDataLoader', async () => {
    const matches = await matchRepo.find({id: {$lte: 5}});
    const partecipants = await Promise.all(
      matches.map(({partecipants}) => partecipants.loadItems())
    );
    const partecipantsDataloader = await Promise.all(
      matches.map(({partecipants}) => dataloader.load(partecipants))
    );
    expect(
      JSON.stringify(
        partecipants.map(users => users.sort((a, b) => a.id - b.id))
      )
    ).toBe(
      JSON.stringify(
        partecipantsDataloader.map(users => users.sort((a, b) => a.id - b.id))
      )
    );
  });

  it('should be faster with EntityDataLoader', async () => {
    const matches = await matchRepo.find({id: {$lte: 5}});
    const start = new Date();
    await Promise.all(
      matches.map(({partecipants}) => partecipants.loadItems())
    );
    const end = new Date();
    const diff = end.getTime() - start.getTime();
    const startDataloader = new Date();
    await Promise.all(
      matches.map(({partecipants}) => dataloader.load(partecipants))
    );
    const endDataloader = new Date();
    const diffDataloader = endDataloader.getTime() - startDataloader.getTime();
    console.log(`W/o dataloader: ${diff} ms`);
    console.log(`W/ dataloader: ${diffDataloader} ms`);
    expect(diffDataloader).toBeLessThan(diff);
  });

  afterAll(async () => {
    orm.close(true);
    await sleep(100);
  });
});
