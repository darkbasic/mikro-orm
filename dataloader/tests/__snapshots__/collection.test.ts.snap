// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backend should retrieve parallel collections with EntityDataLoader 1`] = `
Array [
  Array [
    Object {
      "email": "niccolo.belli@linuxsystems.it",
      "id": 1,
      "name": "Niccolò",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 1,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 3,
          "site": 1,
          "specialty": 1,
        },
      ],
      "password": "1111",
      "picture": null,
      "sex": 0,
      "surname": "Belli",
    },
    Object {
      "email": "mario.rossi@linuxsystems.it",
      "id": 2,
      "name": "Mario",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 1,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 3,
          "site": 1,
          "specialty": 1,
        },
      ],
      "password": "2222",
      "picture": null,
      "sex": 0,
      "surname": "Rossi",
    },
    Object {
      "email": "mattia.ferrari@linuxsystems.it",
      "id": 4,
      "name": "Mattia",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
        Object {
          "creator": 1,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 3,
          "site": 1,
          "specialty": 1,
        },
      ],
      "password": "4444",
      "picture": null,
      "sex": 0,
      "surname": "Ferrari",
    },
    Object {
      "email": "filippo.magi@linuxsystems.it",
      "id": 6,
      "name": "Filippo",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
      ],
      "password": "6666",
      "picture": null,
      "sex": 0,
      "surname": "Magi",
    },
  ],
  Array [
    Object {
      "email": "niccolo.belli@linuxsystems.it",
      "id": 1,
      "name": "Niccolò",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 1,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 3,
          "site": 1,
          "specialty": 1,
        },
      ],
      "password": "1111",
      "picture": null,
      "sex": 0,
      "surname": "Belli",
    },
    Object {
      "email": "mattia.ferrari@linuxsystems.it",
      "id": 4,
      "name": "Mattia",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
        Object {
          "creator": 1,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 3,
          "site": 1,
          "specialty": 1,
        },
      ],
      "password": "4444",
      "picture": null,
      "sex": 0,
      "surname": "Ferrari",
    },
    Object {
      "email": "sara.ricci@linuxsystems.it",
      "id": 7,
      "name": "Sara",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 7,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 5,
          "site": 1,
          "specialty": 2,
        },
      ],
      "password": "7777",
      "picture": null,
      "sex": 1,
      "surname": "Ricci",
    },
    Object {
      "email": "giorgia.esposito@linuxsystems.it",
      "id": 9,
      "name": "Giorgia",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
        Object {
          "creator": 7,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 5,
          "site": 1,
          "specialty": 2,
        },
      ],
      "password": "9999",
      "picture": null,
      "sex": 1,
      "surname": "Esposito",
    },
  ],
  Array [
    Object {
      "email": "mattia.ferrari@linuxsystems.it",
      "id": 4,
      "name": "Mattia",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
        Object {
          "creator": 1,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 3,
          "site": 1,
          "specialty": 1,
        },
      ],
      "password": "4444",
      "picture": null,
      "sex": 0,
      "surname": "Ferrari",
    },
    Object {
      "email": "filippo.magi@linuxsystems.it",
      "id": 6,
      "name": "Filippo",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
      ],
      "password": "6666",
      "picture": null,
      "sex": 0,
      "surname": "Magi",
    },
    Object {
      "email": "giorgia.esposito@linuxsystems.it",
      "id": 9,
      "name": "Giorgia",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
        Object {
          "creator": 7,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 5,
          "site": 1,
          "specialty": 2,
        },
      ],
      "password": "9999",
      "picture": null,
      "sex": 1,
      "surname": "Esposito",
    },
    Object {
      "email": "claudia.recanatini@linuxsystems.it",
      "id": 10,
      "name": "Claudia",
      "partecipatedMatches": Array [
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
        Object {
          "creator": 7,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 5,
          "site": 1,
          "specialty": 2,
        },
      ],
      "password": "0000",
      "picture": null,
      "sex": 1,
      "surname": "Recanatini",
    },
  ],
  Array [
    Object {
      "email": "sara.ricci@linuxsystems.it",
      "id": 7,
      "name": "Sara",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 7,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 5,
          "site": 1,
          "specialty": 2,
        },
      ],
      "password": "7777",
      "picture": null,
      "sex": 1,
      "surname": "Ricci",
    },
    Object {
      "email": "giorgia.esposito@linuxsystems.it",
      "id": 9,
      "name": "Giorgia",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
        Object {
          "creator": 7,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 5,
          "site": 1,
          "specialty": 2,
        },
      ],
      "password": "9999",
      "picture": null,
      "sex": 1,
      "surname": "Esposito",
    },
    Object {
      "email": "claudia.recanatini@linuxsystems.it",
      "id": 10,
      "name": "Claudia",
      "partecipatedMatches": Array [
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
        Object {
          "creator": 7,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 5,
          "site": 1,
          "specialty": 2,
        },
      ],
      "password": "0000",
      "picture": null,
      "sex": 1,
      "surname": "Recanatini",
    },
    Object {
      "email": "lucia.fontana@linuxsystems.it",
      "id": 8,
      "name": "Lucia",
      "partecipatedMatches": Array [
        Object {
          "creator": 7,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 5,
          "site": 1,
          "specialty": 2,
        },
      ],
      "password": "8888",
      "picture": null,
      "sex": 1,
      "surname": "Fontana",
    },
  ],
  Array [
    Object {
      "email": "niccolo.belli@linuxsystems.it",
      "id": 1,
      "name": "Niccolò",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 1,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 3,
          "site": 1,
          "specialty": 1,
        },
      ],
      "password": "1111",
      "picture": null,
      "sex": 0,
      "surname": "Belli",
    },
    Object {
      "email": "mario.rossi@linuxsystems.it",
      "id": 2,
      "name": "Mario",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 1,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 3,
          "site": 1,
          "specialty": 1,
        },
      ],
      "password": "2222",
      "picture": null,
      "sex": 0,
      "surname": "Rossi",
    },
    Object {
      "email": "mattia.ferrari@linuxsystems.it",
      "id": 4,
      "name": "Mattia",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-12T18:00:00.000Z,
          "id": 1,
          "site": 1,
          "specialty": 1,
        },
        Object {
          "creator": 1,
          "date": 2020-05-20T16:30:00.000Z,
          "id": 2,
          "site": 1,
          "specialty": 3,
        },
        Object {
          "creator": 10,
          "date": 2020-05-19T14:00:00.000Z,
          "id": 4,
          "site": 2,
          "specialty": 3,
        },
        Object {
          "creator": 1,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 3,
          "site": 1,
          "specialty": 1,
        },
      ],
      "password": "4444",
      "picture": null,
      "sex": 0,
      "surname": "Ferrari",
    },
    Object {
      "email": "luigi.bianchi@linuxsystems.it",
      "id": 3,
      "name": "Luigi",
      "partecipatedMatches": Array [
        Object {
          "creator": 1,
          "date": 2020-03-15T18:00:00.000Z,
          "id": 3,
          "site": 1,
          "specialty": 1,
        },
      ],
      "password": "3333",
      "picture": null,
      "sex": 0,
      "surname": "Bianchi",
    },
  ],
]
`;
